cmd/gcs/main.go

```diff
@@ -81,6 +85,88 @@ func readMemoryEvents(startTime time.Time, efdFile *os.File, cgName string, thre
 	}
 }
 
+// runWithRestartMonitor starts a command with given args and waits for it to exit. If the
+// command exit code is non-zero the command is restarted with with some back off delay.
+// Any stdout or stderr of the command will be split into lines and written as a log with
+// logrus standard logger.  This function must be called in a separate goroutine.
+func runWithRestartMonitor(arg0 string, args ...string) {
+	backoffSettings := backoff.NewExponentialBackOff()
+	// give after retrying for 10 mins
+	backoffSettings.MaxElapsedTime = time.Minute * 10
+	for {
+		command := exec.Command(arg0, args...)
+		if err := command.Run(); err != nil {
+			logrus.WithFields(logrus.Fields{
+				"error":   err,
+				"command": command.Args,
+			}).Warn("restart monitor: run command returns error")
+		}
+		backOffTime := backoffSettings.NextBackOff()
+		if backOffTime == backoffSettings.Stop {
```

Comment:
`backoff.NewExponentialBackOff()` calls [`Reset()`](https://github.com/cenkalti/backoff#L94) internally which sets the `startTime` field that's used to tell how much time has elapsed since it was instantiated (to determine if it needs to return `backoff.Stop`). If chronyd exited after 10 minutes then we'd never try to restart it as `NextBackOff()` will return `backoff.Stop` on the first call. Can see this with the following:

```go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"time"

	"github.com/cenkalti/backoff/v4"
)

func init() {
	// Need to seed for the rng in backoff.NextBackoff()
	rand.Seed(time.Now().UnixNano())
}

func main() {
	b := backoff.NewExponentialBackOff()
	b.MaxElapsedTime = time.Second * 4
        // Sleep for longer than our max elapsed time 
	time.Sleep(time.Second * 5)
	for {
		backOffTime := b.NextBackOff()
		if backOffTime == b.Stop {
			log.Fatal("oh no")
		}
                // We should never reach the below
		fmt.Println("Goin for a snooze")
		time.Sleep(backOffTime)
	}
}
```

Reply:
Hmm, good catch. I totally missed that. However, now if we are using the backoff package, I don't think there is a simple way of saying "give up if you have already spent 10+ mins trying". I have changed the code to cap the retry interval at 10 mins but now it will keep retrying as long as gcs is running. That should be okay too I guess. Let me know what you think.

Reply:
Yes agreed, don't think there's a simple way to get what we'd wanted there. I think what you changed too is fine!